src/t/d19.clj

```diff
@@ -0,0 +1,219 @@
+(ns t.d19
+  (:gen-class)
+  (:require
+    [clojure.string :as cljstr]
+    [clojure.set :as cljset]
+    [clojure.core.match :refer [match]]))
+
+(def sample
+"r, wr, b, g, bwu, rb, gb, br
+
+brwrr
+bggr
+gbbr
+rrbgbr
+ubwu
+bwurrg
+brgr
+bbrgwb
+")
+(def sol 6)
+
+(defn parse-input
+  [input]
+  (let [[towels patterns] (cljstr/split input #"\n\n")
+        towels (cljstr/split towels #", ")
+        patterns (cljstr/split-lines patterns)]
+    [towels patterns]))
+
+
+(defn p-valid?
+  [p T]
+  (let [psize (count p)
+        t-exact (some (fn [t] (and
+                                (= psize (count t))
+                                (= t p))) T)
+        t-smaller (filter (fn [t] (and
+                                    (< (count t) psize)
+                                    (every? identity (map = t p))))
+                                    T)
+        ]
+;;    (prn [:p p :finish t-exact :t-smaller t-smaller])
+    (or
+      t-exact
+      (some
+        (fn[t] (p-valid? (drop (count t) p) T))
+        t-smaller)
+    ))
+  )
+
+(defn d19p1
+  [input]
+  (let [[T P] (parse-input input)
+        T (map vec T)
+        P (map vec P)
+        p (first P)
+        ]
+    (->> P
+         (keep (fn [p] (p-valid? p T)))
+         count)
+
+    ))
+
+
+
+(defn count-arr-M
+  [T]
+  (memoize
+    (defn count-arr
+      [p]
+      (let [psize (count p)
+            t-forward (filter (fn [t] (and
+                                      (< (count t) psize)
+                                      (every? identity (map = t p))))
+                            T)
+            t-exact (filter (fn[t] (= t p)) T)
+            ]
+        (+ (count t-exact)
+           (apply +
+                  (map
+                    (fn[t] (count-arr (drop (count t) p)))
```

Comment:
Per REPL example above, this recursive call here is calling the non-memoized function that the `t.day19/count-arr` var points to.

Reply:
Or, I guess, it runs the one on lin 103, depending on when `count-arr-M` was called compared to when line 103 was evaluated. Having two `def`s for the same var is also generally frowned upon, as it can make it a bit hard to understand what's going on.

Reply:
mixing names was not part of the plan, I was lacking inspiration for new names and did not think / forgot `defn` would be top level, so I did not see a problem ...